conmon-rs/server/src/fd_socket.rs

```diff
@@ -0,0 +1,268 @@
+//! File descriptor passing via UDS (unix socket).
+//!
+//! ## File descriptor socket protocol
+//!
+//! Connect via a unix `SOCK_SEQPACKET` socket.
+//!
+//! **Request:**
+//! Pass file descriptors via `SCM_RIGHTS` ancillary message.
+//!
+//! `request_id: u32`: Request id selected by client
+//! `num_fds: u8`: Number of file descriptors
+//!
+//! ```txt
+//! u64: (request_id << 32) | num_fds
+//! ```
+//!
+//! **Response:**
+//! ```txt
+//! u64: (request_id << 32) | num_fds
+//! u64: slot for file descriptor 0
+//! u64: slot for file descriptor 1
+//! ...
+//! u64: slot for file descriptor num_fds-1
+//! ```
+//!
+//! On error `num_fds` will be set to 0xffff_ffff and the rest of the message
+//! is an error message string.
+//! ```txt
+//! u64: (request_id << 32) | 0xffff_ffff
+//! ...: error message
+//! ```
+//!
+//! **Close file descriptors:**
+//! Received file descriptors will be closed on disconnect
+//! or with an empty request (`request_id` = 0 and `num_fds` = 0).
+//!
+//! An empty request does **not** receive a response.
+
+use crate::listener::{Listener, SeqpacketListener};
+use anyhow::format_err;
+use anyhow::Result;
+use std::{
+    collections::{hash_map, HashMap},
+    io::{IoSliceMut, Write},
+    mem,
+    num::Wrapping,
+    os::fd::OwnedFd,
+    path::PathBuf,
+    sync::{Arc, Mutex},
+};
+use tokio::{runtime::Handle, sync::Mutex as AsyncMutex, task};
+use tokio_seqpacket::{ancillary::OwnedAncillaryMessage, UnixSeqpacket};
+use tracing::{debug_span, Instrument};
+
+macro_rules! lock {
+    ($x:expr) => {
+        $x.lock().map_err(|e| format_err!("{:#}", e))?
+    };
+}
+
+#[derive(Debug, Default)]
+pub struct FdSocket {
+    server: AsyncMutex<Option<Server>>,
+    state: Mutex<State>,
+}
+
+impl FdSocket {
+    pub async fn start(self: Arc<Self>, path: PathBuf) -> Result<PathBuf> {
+        let mut server = self.server.lock().await;
+        if let Some(server) = server.as_ref() {
+            Ok(server.path.clone())
+        } else {
+            *server = Server::start(path.clone(), self.clone()).await?.into();
+            Ok(path)
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn take(&self, slot: u64) -> Result<OwnedFd> {
+        lock!(self.state).take(slot)
+    }
```

Comment:
why the dead code?

Reply:
`take_all` was previously implemented as a loop over `take`, but refactored to reuse the same lock guard.

Now `take` is unused, but future use cases of the `FdSocket` could work with single file descriptors.

Therefore I think it would be nice to leave this method, so other people can use it in the future without having to understand the internals of the `FdSocket` struct.